pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/PartitionAwareOfflineRoutingTableBuilder.java

```diff
@@ -42,149 +36,143 @@
 import com.linkedin.pinot.common.metadata.segment.SegmentZKMetadata;
 import com.linkedin.pinot.common.response.ServerInstance;
 import com.linkedin.pinot.transport.common.SegmentId;
-import com.linkedin.pinot.transport.common.SegmentIdSet;
 
-public class PartitionAwareOfflineRoutingTableBuilder extends AbstractRoutingTableBuilder {
-  private static final String PARTITION_METADATA_PRUNER = "PartitionZKMetadataPruner";
+/**
+ * Partition aware routing table builder for offline table.
+ *
+ * For an external view change, we need to update the look up table for routing. What we need to do here is to build a
+ * routing look up table that can quickly return the server instance when we are given a segment name and replica id.
+ * Since we assume that the replica group aware segment assignment strategy is used on the controller, we can now prune
+ * servers on the broker. For instance, let's say we have 1 partition, 2 replica groups, 4 servers.
+ *
+ * P0: RG0(server0, server1), RG1(server2, server3)
+ *
+ * Because it is guaranteed that servers of a replica group have a complete set of all segments of the table, we can
+ * pick either RG0(server0, server1) or RG1(server2, server3) for query. This will reduce the number of servers
+ * required for each query to 2 servers from 4 servers.
+ *
+ *
+ * In high level, we need to compute the look up table in the format of (segment_name -> (replica_id -> server_instance))
+ * while we are given the following information.
+ *   - external view: (segment_name -> set(servers))
+ *   - partition to replica group mapping: ((partition_number, replica_id) -> set(servers))
+ *
+ * We go through multiple steps to get the final look up table. The high level algorithm is as follows:
+ *   1. Compute the partition id set by looking at the segment zk metadata and cache metadata when possible.
+ *   2. Build a mapping table of (partition_id -> (server -> replica_id)) using partition set from step 1 and
+ *      the partition to replica group mapping table from property store.
+ *   3. Compute the final routing look up table from external view and the mapping from step 2.
+ *
+ */
+public class PartitionAwareOfflineRoutingTableBuilder extends AbstractPartitionAwareRoutingTableBuilder {
   private static final int TABLE_LEVEL_PARTITION_NUMBER = 0;
```

Comment:
Please add comments to explain the difference between table level vs partition level replica groups